crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+    hash: u64,
+}
+
+impl Hasher for IdentityHasher {
+    #[inline]
+    fn write(&mut self, bytes: &[u8]) {
+        debug_assert_eq!(bytes.len(), 16, "Expected 16 bytes");
+        self.hash = u64::from_le_bytes(bytes[0..8].try_into().unwrap());
+    }
+
+    #[inline]
+    fn write_u64(&mut self, i: u64) {
+        self.hash = i;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.hash
+    }
+}
+type BuildIdentityHasher = BuildHasherDefault<IdentityHasher>;
+
+/// Weighter for the interned string cache. It uses the length of the string plus the size of the
+/// key as the weight. If the string is pinned, its weight is 0, so it will not be evicted from the
+/// cache.
+#[derive(Clone)]
+struct StringWeighter;
+
+impl Weighter<InternedStringId, InternedString> for StringWeighter {
+    fn weight(&self, _key: &InternedStringId, val: &InternedString) -> u64 {
+        if val.1 {
+            0
+        } else {
+            (val.0.len() + size_of::<InternedStringId>()) as u64
+                + size_of::<InternedString>() as u64
+        }
+    }
+}
+
+thread_local! {
+    /// Current step number, used to pin recently created strings in the cache.
+    /// Incremented by each worker on each step.
+    static CURRENT_STEP: RefCell<u64> = const { RefCell::new(0) };
+
+    /// List of pinned strings. On each step, each worker thread unpins and removes
+    /// from the list strings pinned >= 2 steps ago.
+    static PINNED_STRINGS: RefCell<VecDeque<(InternedStringId, (SqlString, u64))>> = const { RefCell::new(VecDeque::new()) };
+}
+
+/// Indexed Z-set that maps interned string IDs back to strings for use by `unintern`.
+///
+/// Each worker maintains its own shard of this Z-set.
+/// Worker 0 collects references to all batches and stores them in this global variable at each step.
+static INTERNED_STRING_BY_ID: LazyLock<
+    RwLock<SpineSnapshot<DynOrdIndexedWSet<DynData, DynData, DynZWeight>>>,
+> = LazyLock::new(|| RwLock::new(empty_by_id()));
+
+/// In-memory cache of interned strings.
+///
+/// This cache serves two purposes.
+/// 1. Fast interned string lookup. The backing store for interned strings is a spine snapshot,
+///    where point lookups can require I/O and are generally more expensive than a hash map lookup
+///    used here.
+/// 2. Provide access to recently interned strings (specifically strings interned during the last
+///    2 steps) that are not yet guaranteed to be in the indexed Z-set.
+///
+// We use the `quick_cache` crate, which in my very limited benchmarking is faster than `lru`,
+// `moka`, and `foyer`.
+#[allow(clippy::type_complexity)]
+static INTERNED_STRING_CACHE: LazyLock<
+    Cache<
+        InternedStringId,
+        InternedString,
+        StringWeighter,
+        BuildIdentityHasher,
+        DefaultLifecycle<InternedStringId, InternedString>,
+    >,
+> = LazyLock::new(|| {
+    Cache::with_options(
+        OptionsBuilder::new()
+            .estimated_items_capacity(CACHE_CAPACITY)
```

Comment:
for testing it would be useful to have the ability to change these
Maybe the cache should have a proper constructor and a Default trait.

Reply:
That's a good point, I'll figure out a way to do this.